batcher/sweep_batch.go

```diff
@@ -0,0 +1,1330 @@
+		}
+
+		// To be sure that we're good, parse and validate that the combined
+		// signature is indeed valid for the sig hash and the internal pubkey.
+		err = b.verifySchnorrSig(
+			htlcScript.TaprootKey, sigHash, finalSig,
+		)
+		if err != nil {
+			return err
+		}
+
+		packet.UnsignedTx.TxIn[i].Witness = wire.TxWitness{
+			finalSig,
+		}
+	}
+
+	return nil
+}
+
+// updateRbfRate updates the fee rate we should use for the new batch
+// transaction. This fee rate does not guarantee RBF success, but the continuous
+// increase leads to an eventual successful RBF replacement.
+func (b *Batch) updateRbfRate(ctx context.Context) error {
+	// If the feeRate is unset then we never published before, so we
+	// retrieve the fee estimate from our wallet.
+	if b.RbfCache.FeeRate == 0 {
+		b.log.Infof("initializing rbf fee rate for conf target=%v", b.Cfg.BatchConfTarget)
+		rate, err := b.wallet.EstimateFeeRate(
+			ctx, b.Cfg.BatchConfTarget,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Set the initial value for our fee rate.
+		b.RbfCache.FeeRate = rate
+	} else {
+		// Bump the fee rate by the configured step.
+		b.RbfCache.FeeRate += defaultFeeRateStep
+		b.RbfCache.LastHeight = b.currentHeight
+		return b.persist(ctx)
+	}
+
+	return nil
+}
+
+// monitorBlockHeight monitors the current block height.
+func (b *Batch) monitorBlockHeight(ctx context.Context) {
+	blockChan, errChan, err := b.chainNotifier.RegisterBlockEpochNtfn(
+		ctx,
+	)
+	if err != nil {
+		b.errChan <- err
+		return
+	}
+
+	for {
+		select {
+		case height := <-blockChan:
+			select {
+			case b.blockEpochChan <- height:
+			case <-ctx.Done():
+				return
+			}
+
+		case err := <-errChan:
+			select {
+			case b.errChan <- err:
+			case <-ctx.Done():
+				return
+			}
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// monitorSpend monitors the primary sweep's outpoint for spends. The reason we
+// monitor the primary sweep's outpoint is because the primary sweep was the
+// first sweep that entered this batch, therefore it is present in all the
+// versions of the batch transaction. This means that even if an older version
+// of the batch transaction gets confirmed, due to the uncertainty of RBF
+// replacements and network propagation, we can always detect the transaction.
+func (b *Batch) monitorSpend(ctx context.Context) {
+	// If no primary sweep exists the batch is empty and there is no spend
+	// to monitor.
+	if b.PrimarySweepID == lntypes.ZeroHash {
+		b.log.Warnf("can't monitor batch spend, no primary sweep")
+		return
+	}
+
+	// Generate this routine's nonce. This is used to signal to the previous
+	// routine that a new routine is now monitoring the spend.
+	randValue, err := crand.Int(crand.Reader, big.NewInt(math.MaxInt32))
+	if err != nil {
+		b.errChan <- err
+		return
+	}
+
+	nonce := int32(randValue.Int64())
+	// Send the nonce to the refresh channel.
+	b.spendRefreshChan <- nonce
+
+	// Get the primary sweep.
+	b.sweepsLock.Lock()
```

Comment:
Instead of locking here, the function could just receive the primary sweep's pkscript, etc as arguments. That way we'd  also make sure that we don't mistakenly call the function if the primary sweep is not yet set.

Reply:
this function is called in a goroutine, so then all the callers would have to lock before calling to acquire the values that would be placed in the func args